polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2396,6 +2396,287 @@ ValueCategory MLIRScanner::EmitPromotedScalarExpr(Expr *E,
+  // non-scalar types. Remove when we actually call operatorOP member functions.
+  if (!(SrcType.isNull() || SrcType.isCanonical()) ||
+      !(DstType.isNull() || DstType.isCanonical())) {
+    return Src;
+  }
+
+  // TODO: Handle fixed points here when supported.
+  // TODO: Take into account scalar conversion options.
+
+  assert(!SrcType->isFixedPointType() &&
+         "Not handling conversion from fixed point types");
+
+  assert(!DstType->isFixedPointType() &&
+         "Not handling conversion to fixed point types");
+
+  mlirclang::CodeGen::CodeGenTypes &CGTypes = Glob.getTypes();
+
+  SrcType = Glob.getCGM().getContext().getCanonicalType(SrcType);
+  DstType = Glob.getCGM().getContext().getCanonicalType(DstType);
+  if (SrcType == DstType)
+    return Src;
+  if (DstType->isVoidType())
+    return nullptr;
+
+  auto SrcTy = Src.val.getType();
+
+  if (DstType->isBooleanType()) {
+    llvm::WithColor::warning() << "Performing conversion to bool as if it were "
+                                  "a regular integer type\n";
+  }
+
+  mlir::Type DstTy = CGTypes.getMLIRType(DstType);
+
+  // Cast from half through float if half isn't a native type.
+  const auto &CGM = Glob.getCGM();
+  if (SrcType->isHalfType() && !CGM.getLangOpts().NativeHalfType) {
+    // Cast to FP using the intrinsic if the half type itself isn't supported.
+    if (DstTy.isa<FloatType>()) {
+      if (CGM.getContext().getTargetInfo().useFP16ConversionIntrinsics())
+        llvm::WithColor::warning() << "Should call convert_from_fp16 intrinsic "
+                                      "to perfom this conversion\n";
+    } else {
+      // Cast to other types through float, using either the intrinsic or FPExt,
+      // depending on whether the half type itself is supported
+      // (as opposed to operations on half, available with NativeHalfType).
+      if (CGM.getContext().getTargetInfo().useFP16ConversionIntrinsics())
+        llvm::WithColor::warning() << "Should call convert_from_fp16 intrinsic "
+                                      "to perfom this conversion\n";
+      Src = Src.FPExt(builder, builder.getF32Type());
+      SrcType = CGM.getContext().FloatTy;
+      SrcTy = builder.getF32Type();
+    }
+  }
+
+  // Ignore conversions like int -> uint.
+  if (SrcTy == DstTy) {
+    llvm::WithColor::warning()
+        << "Not emitting implicit integer sign change checks\n";
+
+    return Src;
+  }
+
+  // Handle pointer conversions next: pointers can only be converted to/from
+  // other pointers and integers. Check for pointer types in terms of LLVM, as
+  // some native types (like Obj-C id) may map to a pointer type.
+  assert(!(DstTy.isa<MemRefType>() || SrcTy.isa<MemRefType>()) &&
+         "Not implemented yet");
+
+  // A scalar can be splatted to an extended vector of the same element type
+  assert(!(DstType->isExtVectorType() && !SrcType->isVectorType()) &&
+         "Not implemented yet");
+
+  if (SrcType->isMatrixType() && DstType->isMatrixType())
```

Comment:
Should we mark this case as not implemented yet too, as `EmitScalarCast` would immediately assert for this case?